File: sample-objc/ARGEARSample/ARGearHelpers/ARGCamera/Permission.m
//
//  Permission.m
//  ARGEARSample
//
//  Created by Jihye on 19/08/2019.
//  Copyright © 2019 Seerslab. All rights reserved.
//

#import "Permission.h"
#import <Photos/Photos.h>

@implementation Permission

- (void)allowCameraAction:(void(^)(void))completionHandler {
    AVAuthorizationStatus videoStatus = [AVCaptureDevice authorizationStatusForMediaType:AVMediaTypeVideo];
    if (videoStatus == AVAuthorizationStatusDenied) {
        dispatch_async( dispatch_get_main_queue(), ^{
            completionHandler();
        });
    } else {
        [AVCaptureDevice requestAccessForMediaType:AVMediaTypeVideo completionHandler:^(BOOL granted) {
            dispatch_async( dispatch_get_main_queue(), ^{
                completionHandler();
            });
        }];
    }
}

- (void)allowLibraryAction:(void(^)(void))completionHandler {
    PHAuthorizationStatus libraryStatus = [PHPhotoLibrary authorizationStatus];
    if (libraryStatus == PHAuthorizationStatusDenied) {
        dispatch_async(dispatch_get_main_queue(), ^{
            completionHandler();
        });
    } else {
        [PHPhotoLibrary requestAuthorization:^(PHAuthorizationStatus status) {
            dispatch_async( dispatch_get_main_queue(), ^{
                BOOL authorizeStatus = (status == PHAuthorizationStatusAuthorized) ? YES : NO;
                completionHandler();
            });
        }];
    }
}

- (void)allowMicAction:(void(^)(void))completionHandler {
    AVAuthorizationStatus audioStatus = [AVCaptureDevice authorizationStatusForMediaType:AVMediaTypeAudio];
    if (audioStatus == AVAuthorizationStatusDenied) {
        dispatch_async( dispatch_get_main_queue(), ^{
            completionHandler();
        });
    } else {
        [AVCaptureDevice requestAccessForMediaType:AVMediaTypeAudio completionHandler:^(BOOL granted) {
            dispatch_async( dispatch_get_main_queue(), ^{
                completionHandler();
            });
        }];
    }
}

- (void)permissionAllAllowAction:(PermissionEndBlock)endedBlock {

    [self allowCameraAction:^{
        [self allowLibraryAction:^{
            [self allowMicAction:^{
                [endedBlock invoke];
            }];
        }];
    }];
}

- (PermissionLevel)getPermissionLevel {
    AVAuthorizationStatus videoStatus = [AVCaptureDevice authorizationStatusForMediaType:AVMediaTypeVideo];
    AVAuthorizationStatus audioStatus = [AVCaptureDevice authorizationStatusForMediaType:AVMediaTypeAudio];
    PHAuthorizationStatus libraryStatus = [PHPhotoLibrary authorizationStatus];
    
    BOOL videoDetermined = (videoStatus == AVAuthorizationStatusNotDetermined) ? YES : NO;
    BOOL audioDetermined = (audioStatus == AVAuthorizationStatusNotDetermined) ? YES : NO;
    BOOL libraryDetermined = (libraryStatus == AVAuthorizationStatusNotDetermined) ? YES : NO;
    
    BOOL videoAuthorized = (videoStatus == AVAuthorizationStatusAuthorized) ? YES : NO;
    BOOL audioAuthorized = (audioStatus == AVAuthorizationStatusAuthorized) ? YES : NO;
    BOOL libraryAuthorized = (libraryStatus == AVAuthorizationStatusAuthorized) ? YES : NO;
    
    if (videoDetermined || audioDetermined || libraryDetermined) {
        return PermissionLevelNone;
    }
    
    if (videoAuthorized && audioAuthorized && libraryAuthorized) {
        return PermissionLevelGranted;
    }
    
    return PermissionLevelRestricted;
}

- (void)openSettings {
    
    if (@available(iOS 10.0, *)) {
        [[UIApplication sharedApplication] openURL:[NSURL URLWithString:UIApplicationOpenSettingsURLString] options:@{} completionHandler:nil];
    } else {
        // Fallback on earlier versions (under iOS 10)
        [[UIApplication sharedApplication] openURL:[NSURL URLWithString:UIApplicationOpenSettingsURLString]];
    }
}

@end

File: sample-swift/ARGearSample/ARGearHelpers/ARGCamera/Permission.swift
//
//  Permission.swift
//  ARGearSample
//
//  Created by Jihye on 2020/01/21.
//  Copyright © 2020 Seerslab. All rights reserved.
//

import UIKit
import Photos

enum PermissionLevel : Int {
    case None       = 0
    case Granted    = 1
    case Restricted = 2
}

typealias PermissionCheckComplete = () -> Void
typealias PermissionCheckHandler = (PermissionLevel) -> Void

class Permission: NSObject {
    
    var grantedHandler: PermissionCheckComplete?
    
    func allowCamera(_ permissionCheckComplete: @escaping PermissionCheckComplete) {
        let cameraStatus = AVCaptureDevice.authorizationStatus(for: .video)
        if cameraStatus == .denied {
            permissionCheckComplete()
        } else {
            AVCaptureDevice.requestAccess(for: .video) { granted in
                permissionCheckComplete()
            }
        }
    }
    
    func allowLibrary(_ permissionCheckComplete: @escaping PermissionCheckComplete) {
        let libraryStatus = PHPhotoLibrary.authorizationStatus()
        if libraryStatus == .denied {
            permissionCheckComplete()
        } else {
            PHPhotoLibrary.requestAuthorization { status in
                permissionCheckComplete()
            }
        }
    }
    
    func allowMic(_ permissionCheckComplete: @escaping PermissionCheckComplete) {
        let micStatus = AVCaptureDevice.authorizationStatus(for: .audio)
        if micStatus == .denied {
            permissionCheckComplete()
        } else {
            AVCaptureDevice.requestAccess(for: .audio) { granted in
                permissionCheckComplete()
            }
        }
    }
    
    func allAllow(_ permissionCheckHandler: @escaping PermissionCheckHandler) {
        self.allowCamera {
            self.allowLibrary {
                self.allowMic {
                    permissionCheckHandler(self.getPermissionLevel())
                }
            }
        }
    }
    
    func getPermissionLevel() -> PermissionLevel {
        let cameraStatus = AVCaptureDevice.authorizationStatus(for: .video)
        let micStatus = AVCaptureDevice.authorizationStatus(for: .audio)
        let libraryStatus = PHPhotoLibrary.authorizationStatus()
        
        let cameraDetermined = (cameraStatus == .notDetermined) ? true : false
        let micDetermined = (micStatus == .notDetermined) ? true : false
        let libraryDetermined = (libraryStatus == .notDetermined) ? true : false
        
        let cameraAuthorized = (cameraStatus == .authorized) ? true : false
        let micAuthorized = (micStatus == .authorized) ? true : false
        let libraryAuthorized = (libraryStatus == .authorized) ? true : false
        
        if cameraDetermined || micDetermined || libraryDetermined {
            return .None
        }
        
        if cameraAuthorized && micAuthorized && libraryAuthorized {
            return .Granted
        }
        
        return .Restricted
    }
    
    func openSettings() {
        guard let settingsURL = URL(string: UIApplication.openSettingsURLString)
            else { return }
        
        if #available(iOS 10.0, *) {
            UIApplication.shared.open(settingsURL, options: [:], completionHandler: nil)
        } else {
            UIApplication.shared.openURL(settingsURL)
        }
    }
}
